zdm/grid.py

```diff
@@ -0,0 +1,672 @@
+        return FRBparams,pwb
+        
+
+    def update(self, vparams:dict):
+        """[summary]
+        
+        Hierarchy:
+        Each indent corresponds to one 'level'.
+        This is used in the program control below
+        to dictate how far each tree should proceed
+        in calculation.
+        Direct variable inputs are always listed first
+        We see that sfr evolution and dm smearing
+        lie just before the pdv step
+        Hence, we deal with these first, and
+        calc rates as a final step regardless
+        of what else has changed.
+        
+        calc_rates:
+            calc_pdv
+                Emin
+                Emax
+                gamma
+                calc_thresholds
+                    F0
+                    alpha
+                    bandwidth
+            set_evolution
+                sfr_n
+            
+            smear_grid:
+                grid
+                mask
+                    dmx_params
+        
+        Note that the grid is independent of the constant C (trivial dependence)
+
+        Args:
+            vparams (dict): [description]
+        """
+        # Init
+        smear_mask, smear_dm, calc_pdv, set_evol = False, False, False, False
+        new_sfr_smear, new_pdv_smear, calc_thresh = False, False, False
+
+        # Cosmology -- Only H0 so far
+        if self.chk_upd_param('H0', vparams, update=True):
+            cos.set_cosmology(self.state)
+            zDMgrid, zvals,dmvals=misc_functions.get_zdm_grid(
+                self.state, new=True,plot=False,method='analytic')
+            # TODO -- Check zvals and dmvals haven't changed!
+            self.pass_grid(zDMgrid,zvals,dmvals)
+            # The rest
+            smear_mask = True
+            smear_dm = True
+            calc_thresh = True
+            calc_pdv = True
+            set_evol = True
+            new_sfr_smear = True
+
+        # Mask?
+        # IT IS IMPORTANT TO USE np.any so that each item is executed!!
+        if np.any([self.chk_upd_param('lmean', vparams, update=True), 
+            self.chk_upd_param('lsigma', vparams, update=True),
+            smear_mask]):
+            self.smear=pcosmic.get_dm_mask(
+                self.dmvals,(self.state.host.lmean,
+                             self.state.host.lsigma),
+                self.zvals)
+            smear_dm = True
+
+        # Smear?
+        if smear_dm:
+            self.smear_dm(self.smear)
+
+        # SFR?
+        if self.chk_upd_param('sfr_n', vparams, update=True):
+            set_evol = True
+            new_sfr_smear=True  # True for either alpha_method
+        if self.chk_upd_param('alpha', vparams, update=True):
+            set_evol = True
+            calc_pdv = True
+            new_pdv_smear=True
+            if self.state.FRBdemo.alpha_method == 0:
+                calc_thresh = True
+            elif self.state.FRBdemo.alpha_method == 1:
+                new_sfr_smear=True
+        if set_evol:
```

Comment:
I think we should re-order this routine into two halves (it is almost there anyway):
#1: perform all the checks for changes parameters, and set changes to True/False accordingly
#2: run through all the IF (changes) THEN (calculate) steps
We are close, but there are a couple of tings out-of-order below

Reply:
done